modin/backends/pandas/parsers.py

```diff
@@ -153,15 +153,31 @@ def get_dtypes(cls, dtypes_ids):
             pandas.Series where index is columns names and values are
             columns dtypes.
         """
+        # each element in `partitions_dtypes` is a Series, where column names are
+        # used as index and types of columns for different partitions are used as values
         partitions_dtypes = cls.materialize(dtypes_ids)
         if all([len(dtype) == 0 for dtype in partitions_dtypes]):
             return None
-        return (
-            pandas.concat(partitions_dtypes, axis=1)
-            .apply(lambda row: find_common_type_cat(row.values), axis=1)
-            .squeeze(axis=0)
+
+        # concat all elements of `partitions_dtypes` and find common dtype
+        # for each of the column among all partitions
+        combined_dtypes = pandas.concat(partitions_dtypes, axis=1).apply(
+            lambda row: find_common_type_cat(row.values),
+            axis=1,
         )
 
+        for part_dtype in partitions_dtypes:
```

Comment:
This will be really expensive for really large clusters and for very wide data.

What about a fastpath for all equal data types:

```python
combined_part_dtypes = pandas.concat(partitions_dtypes, axis=1)
if combined_part_dtypes.eq(combined_part_dtypes.iloc[0], axis=1).all(axis=None):
    combined_dtypes = combined_part_dtypes.iloc[0]
else:
    ErrorMessage.missmatch_with_pandas(...)

```

Does this make sense?

Reply:
Yes, for very wide data and large number of partitions it potentially can be expensive. Implemented suggestion with couple corrections, so now common for all partitions dtypes `combined_part_dtypes.iloc[0]` is calculated only if heterogeneous data is detected.